File: backend/app/schemas/project.py
import uuid
from datetime import datetime
from typing import List, Optional
from pydantic import BaseModel, Field

class ProjectBase(BaseModel):
    """
    Base shared properties for projects.
    """
    name: str = Field(..., min_length=1, max_length=255)
    slug: str = Field(..., min_length=1, max_length=255)
    description: Optional[str] = Field(None, max_length=1024)
    icon: Optional[str] = Field(None, max_length=50)
    cover_image: Optional[str] = Field(None, max_length=1024)
    color: Optional[str] = Field(None, max_length=50)
    status: str = Field("Pending", max_length=50)
    priority: str = Field("Medium", max_length=50)
    progress: int = Field(0, ge=0, le=100)
    is_favorite: bool = Field(False)
    is_pinned: bool = Field(False)
    is_archived: bool = Field(False)
    workspace_id: Optional[uuid.UUID] = None
    visibility: str = Field("Workspace", max_length=50)
    due_date: Optional[datetime] = None
    kanban_columns: Optional[List[dict]] = None

class ProjectCreate(BaseModel):
    """
    Data structure for project creation.
    """
    name: str = Field(..., min_length=1, max_length=255)
    slug: Optional[str] = Field(None, max_length=255)
    description: Optional[str] = Field(None, max_length=1024)
    icon: Optional[str] = Field(None, max_length=50)
    cover_image: Optional[str] = Field(None, max_length=1024)
    color: Optional[str] = Field(None, max_length=50)
    status: Optional[str] = Field("Pending", max_length=50)
    priority: Optional[str] = Field("Medium", max_length=50)
    workspace_id: Optional[uuid.UUID] = None
    visibility: Optional[str] = Field("Workspace", max_length=50)
    due_date: Optional[datetime] = None
    kanban_columns: Optional[List[dict]] = None

class ProjectUpdate(BaseModel):
    """
    Data structure for project editing. All fields are optional.
    """
    name: Optional[str] = Field(None, min_length=1, max_length=255)
    slug: Optional[str] = Field(None, max_length=255)
    description: Optional[str] = Field(None, max_length=1024)
    icon: Optional[str] = Field(None, max_length=50)
    cover_image: Optional[str] = Field(None, max_length=1024)
    color: Optional[str] = Field(None, max_length=50)
    status: Optional[str] = Field(None, max_length=50)
    priority: Optional[str] = Field(None, max_length=50)
    progress: Optional[int] = Field(None, ge=0, le=100)
    is_favorite: Optional[bool] = Field(None)
    is_pinned: Optional[bool] = Field(None)
    is_archived: Optional[bool] = Field(None)
    workspace_id: Optional[uuid.UUID] = None
    visibility: Optional[str] = Field(None, max_length=50)
    due_date: Optional[datetime] = None
    kanban_columns: Optional[List[dict]] = None

class ProjectMember(BaseModel):
    """
    Simple schema for user memberships in project teams.
    """
    id: uuid.UUID
    full_name: str
    profile_image: Optional[str] = None

    class Config:
        from_attributes = True

class ProjectResponse(ProjectBase):
    """
    Database representation schema returned to clients.
    """
    id: uuid.UUID
    owner_id: uuid.UUID
    created_at: datetime
    updated_at: datetime
    members: List[ProjectMember] = []

    class Config:
        from_attributes = True

File: frontend/services/project.ts
import { api } from './api'

export interface ProjectCreateInput {
  name: string
  slug?: string
  description?: string
  icon?: string
  cover_image?: string
  color?: string
  status?: string
  priority?: string
  visibility?: string
  workspace_id?: string
}

export interface ProjectResponse {
  id: string
  name: string
  slug: string
  description: string | null
  icon: string | null
  cover_image: string | null
  color: string | null
  status: string
  priority: string
  progress: number
  is_favorite: boolean
  is_pinned: boolean
  is_archived: boolean
  owner_id: string
  workspace_id: string | null
  visibility: string
  created_at: string
  updated_at: string
  members: Array<{
    id: string
    full_name: string
    profile_image: string | null
  }>
}

export interface ProjectListParams {
  search?: string
  is_archived?: boolean
  status_filter?: string
  priority?: string
  visibility?: string
  is_favorite?: boolean
  is_pinned?: boolean
  owner_id?: string
  sort_by?: string
}

export interface ProjectStatisticsResponse {
  total_projects: number
  active_projects: number
  archived_projects: number
  pinned_projects: number
  favorite_projects: number
  average_progress: number
}

export const projectService = {
  /**
   * Load user projects list with full filtering parameter scopes.
   */
  async getProjects(params?: ProjectListParams): Promise<ProjectResponse[]> {
    const response = await api.get<ProjectResponse[]>('/projects', { params })
    return response.data
  },

  /**
   * Fetch the most recently modified projects.
   */
  async getRecentProjects(): Promise<ProjectResponse[]> {
    const response = await api.get<ProjectResponse[]>('/projects/recent')
    return response.data
  },

  /**
   * Retrieve project details.
   */
  async getProjectById(id: string): Promise<ProjectResponse> {
    const response = await api.get<ProjectResponse>(`/projects/${id}`)
    return response.data
  },

  /**
   * Create a new project.
   */
  async createProject(data: ProjectCreateInput): Promise<ProjectResponse> {
    const response = await api.post<ProjectResponse>('/projects', data)
    return response.data
  },

  /**
   * Edit existing project fields.
   */
  async updateProject(id: string, data: Partial<ProjectResponse>): Promise<ProjectResponse> {
    const response = await api.put<ProjectResponse>(`/projects/${id}`, data)
    return response.data
  },

  /**
   * Archive project.
   */
  async archiveProject(id: string): Promise<ProjectResponse> {
    const response = await api.patch<ProjectResponse>(`/projects/archive/${id}`)
    return response.data
  },

  /**
   * Restore archived project.
   */
  async restoreProject(id: string): Promise<ProjectResponse> {
    const response = await api.patch<ProjectResponse>(`/projects/restore/${id}`)
    return response.data
  },

  /**
   * Toggle favorite project flag.
   */
  async favoriteProject(id: string): Promise<ProjectResponse> {
    const response = await api.patch<ProjectResponse>(`/projects/favorite/${id}`)
    return response.data
  },

  /**
   * Toggle pinned project flag.
   */
  async pinProject(id: string): Promise<ProjectResponse> {
    const response = await api.patch<ProjectResponse>(`/projects/pin/${id}`)
    return response.data
  },

  /**
   * Update progress value.
   */
  async updateProgress(id: string, progress: number): Promise<ProjectResponse> {
    const response = await api.patch<ProjectResponse>(`/projects/progress/${id}?progress=${progress}`)
    return response.data
  },

  /**
   * Load workspace project metrics/statistics.
   */
  async getStatistics(): Promise<ProjectStatisticsResponse> {
    const response = await api.get<ProjectStatisticsResponse>('/projects/statistics')
    return response.data
  },

  /**
   * Delete a project.
   */
  async deleteProject(id: string): Promise<void> {
    await api.delete(`/projects/${id}`)
  },
}
